# 1 - Zipcodes.
## Making the zipcode dictionary.

### We will us the USZipCode python package

In [52]:
import pandas as pd
from uszipcode import SearchEngine, SimpleZipcode, Zipcode

### US Zip Code Query - massachusetts_usazip

In [70]:
search = SearchEngine() 
#SearchEngine(simple_zipcode=False) #This query will return more data
results = search.by_state("Massachusetts",returns=0) #default returns 5, 0 returns all

In [73]:
massachusetts_usazip = []

for entry in results:
    temp = entry.to_dict()
    massachusetts_usazip.append(temp)

massachusetts_usazip = pd.DataFrame(massachusetts_zipcodes)
massachusetts_usazip.columns = map(str.capitalize, massachusetts_usazip.columns)

In [74]:
massachusetts_usazip.head()

Zipcode Zipcode_type   Major_city Post_office_city  \
0   01001     Standard       Agawam       Agawam, MA   
1   01002     Standard      Amherst      Amherst, MA   
2   01005     Standard        Barre        Barre, MA   
3   01007     Standard  Belchertown  Belchertown, MA   
4   01008     Standard    Blandford    Blandford, MA   

             Common_city_list            County State    Lat    Lng Timezone  \
0                    [Agawam]    Hampden County    MA  42.07 -72.63  Eastern   
1  [Amherst, Cushman, Pelham]  Hampshire County    MA  42.38 -72.52  Eastern   
2                     [Barre]  Worcester County    MA  42.42 -72.12  Eastern   
3               [Belchertown]  Hampshire County    MA  42.30 -72.40  Eastern   
4                 [Blandford]    Hampden County    MA  42.20 -73.00  Eastern   

   ...  Land_area_in_sqmi Water_area_in_sqmi  Housing_units  \
0  ...              11.44               0.86         7557.0   
1  ...              55.04               1.65        10388.0   
2  ...              44.24               0.26         2044.0   
3  ...              52.64               2.68         5839.0   
4  ...              53.80               1.96          586.0   

   Occupied_housing_units  Median_home_value  Median_household_income  \
0                  7215.0           213000.0                  58733.0   
1                  9910.0           338900.0                  54422.0   
2                  1904.0           208500.0                  68644.0   
3                  5595.0           260000.0                  71875.0   
4                   503.0           247200.0                  71635.0   

   Bounds_west  Bounds_east  Bounds_north  Bounds_south  
0   -72.667902   -72.582535     42.100467     42.030795  
1   -72.546776   -72.355041     42.437947     42.301437  
2   -72.205174   -72.007388     42.484473     42.356423  
3   -72.472287   -72.331642     42.358762     42.185812  
4   -73.034916   -72.872180     42.251340     42.113028  

[5 rows x 24 columns]

### Write out the Massachusetts usazip + Demographic information to csv

In [76]:
# write out the massachusettes zipcodes we will be working with
massachusetts_usazip.to_csv('~/Desktop/new_insights/data_files/massachusetts_usazip.csv', index=False)
massachusetts_usazip.head()

Zipcode Zipcode_type   Major_city Post_office_city  \
0   01001     Standard       Agawam       Agawam, MA   
1   01002     Standard      Amherst      Amherst, MA   
2   01005     Standard        Barre        Barre, MA   
3   01007     Standard  Belchertown  Belchertown, MA   
4   01008     Standard    Blandford    Blandford, MA   

             Common_city_list            County State    Lat    Lng Timezone  \
0                    [Agawam]    Hampden County    MA  42.07 -72.63  Eastern   
1  [Amherst, Cushman, Pelham]  Hampshire County    MA  42.38 -72.52  Eastern   
2                     [Barre]  Worcester County    MA  42.42 -72.12  Eastern   
3               [Belchertown]  Hampshire County    MA  42.30 -72.40  Eastern   
4                 [Blandford]    Hampden County    MA  42.20 -73.00  Eastern   

   ...  Land_area_in_sqmi Water_area_in_sqmi  Housing_units  \
0  ...              11.44               0.86         7557.0   
1  ...              55.04               1.65        10388.0   
2  ...              44.24               0.26         2044.0   
3  ...              52.64               2.68         5839.0   
4  ...              53.80               1.96          586.0   

   Occupied_housing_units  Median_home_value  Median_household_income  \
0                  7215.0           213000.0                  58733.0   
1                  9910.0           338900.0                  54422.0   
2                  1904.0           208500.0                  68644.0   
3                  5595.0           260000.0                  71875.0   
4                   503.0           247200.0                  71635.0   

   Bounds_west  Bounds_east  Bounds_north  Bounds_south  
0   -72.667902   -72.582535     42.100467     42.030795  
1   -72.546776   -72.355041     42.437947     42.301437  
2   -72.205174   -72.007388     42.484473     42.356423  
3   -72.472287   -72.331642     42.358762     42.185812  
4   -73.034916   -72.872180     42.251340     42.113028  

[5 rows x 24 columns]

### Is the python USZIP package a good dataset?
- We can look at the county level.
- Are the zip codes in the dataset (and the demographic data) capturing the expectations for the counties?

In [77]:
countyValues = massachusetts_usazip['County'].unique()
countyValues = countyValues.tolist()
countyValues = sorted(countyValues)

for county in countyValues:
    print(county)
    temp = massachusetts_zipcodes[massachusetts_zipcodes['County']==county]
    print("land area (mi2):",temp['Land_area_in_sqmi'].sum())
    print("population:", temp['Population'].sum())
    print()

Barnstable County
land area (mi2): 367.31999999999994
population: 212674.0

Berkshire County
land area (mi2): 868.03
population: 128093.0

Bristol County
land area (mi2): 551.5500000000001
population: 546665.0

Dukes County
land area (mi2): 65.06
population: 13735.0

Essex County
land area (mi2): 490.41
population: 743207.0

Franklin County
land area (mi2): 674.69
population: 70277.0

Hampden County
land area (mi2): 606.97
population: 461096.0

Hampshire County
land area (mi2): 533.67
population: 147976.0

Middlesex County
land area (mi2): 814.62
population: 1494025.0

Nantucket County
land area (mi2): 39.13
population: 9852.0

Norfolk County
land area (mi2): 394.78999999999996
population: 683167.0

Plymouth County
land area (mi2): 654.5999999999999
population: 490419.0

Suffolk County
land area (mi2): 56.11
population: 716794.0

Worcester County
land area (mi2): 1437.52
population: 791155.0



In [ ]:
######## from wikipedia ##########

# barnstable  215,888    396mi^2 #
# berkshire   131,219    931mi^2 #
# bristol     548,285    556mi^2 #
# dukes        16,535    104mi^2 #  vs 65mi^2 <- Dukes county is a bit weird, lots of islands and water.
# essex       743,159    498mi^2 #
# franklin     71,372    702mi^2 #
# hampden     463,490    618mi^2 #
# hampshire   158,080    529mi^2 #
# middlesex 1,503,085    824mi^2 #
# nantucket    10,172     48mi^2 #
# norfolk     670,850    400mi^2 #
# plymouth    494,919    661mi^2 #
# suffolk     722,023     58mi^2 #
# worcester   798,552   1513mi^2 #

- overall pretty good, might want to look at Dukes County in the future

### We want to represent these zipcodes in reference to a list of 'target cities'. The target cities come from the city level data from the Massachusettes State Government
- webpage here...

In [60]:
massachusetts_county_overdoses = pd.read_csv('~/Desktop/new_insights/data_files/massachusetts_county_overdoses.csv', dtype = 'str', index_col= False)
massachusetts_city_overdoses = pd.read_csv('~/Desktop/new_insights/data_files/massachusetts_city_overdoses.csv', dtype = 'str', index_col= False)

In [80]:
target_cities = massachusetts_city_overdoses['City']

### How does the USAZipcode database differ from our 'target cities' in Massachusettes?

In [82]:
print("There are", str(len(massachusetts_usazip['Major_city'])), "zipcodes in the massachusetts usazip database")

There are 496 zipcodes in the massachusetts usazip database


In [83]:
print("There are", str(len(massachusetts_zipcodes['Major_city'].unique())), "unique 'cities' in the massachusetts usazip database")

There are 418 unique 'cities' in the massachusetts usazip database


In [84]:
print("There are", str(len(target_cities)), "cities in the target cities list (ie MA overdose city data)")

There are 352 cities in the target cities list (ie MA overdose city data)


### I went through these discrepencies and assigned 'target' cities to zipcodes where it was clear

- renaming
- assessing based on "common city name"

In [97]:
target_cities_notin_massachusetts_usazip = []

for city in target_cities:
    if city in list(massachusetts_usazip['Major_city']):
        print(city)
    if city not in list(massachusetts_usazip['Major_city']):
        print('>', city)
        target_cities_notin_massachusetts_usazip.append(city)

Abington
Acton
Acushnet
Adams
Agawam
> Alford
Amesbury
Amherst
Andover
> Aquinnah
Arlington
Ashburnham
Ashby
Ashfield
Ashland
Athol
Attleboro
Auburn
Avon
Ayer
Barnstable
Barre
Becket
Bedford
Belchertown
Bellingham
Belmont
Berkley
Berlin
Bernardston
Beverly
Billerica
Blackstone
Blandford
Bolton
Boston
> Bourne
Boxborough
Boxford
Boylston
Braintree
Brewster
Bridgewater
Brimfield
Brockton
Brookfield
Brookline
Buckland
Burlington
Cambridge
Canton
Carlisle
Carver
Charlemont
Charlton
Chatham
Chelmsford
Chelsea
Cheshire
Chester
Chesterfield
Chicopee
Chilmark
> Clarksburg
Clinton
Cohasset
Colrain
Concord
Conway
Cummington
Dalton
Danvers
> Dartmouth
Dedham
Deerfield
Dennis
Dighton
Douglas
Dover
Dracut
Dudley
Dunstable
Duxbury
East Bridgewater
East Brookfield
East Longmeadow
Eastham
Easthampton
> Easton
Edgartown
> Egremont
Erving
Essex
Everett
Fairhaven
Fall River
Falmouth
Fitchburg
> Florida
> Foxborough
Framingham
Franklin
> Freetown
Gardner
Georgetown
Gill
Gloucester
Goshen
> Gosnold
Grafton

In [99]:
print('There are', str(len(target_cities_notin_massachusetts_usazip)), 'target cities missing from massachusetts usazip list.')       

There are 41 target cities missing from massachusetts usazip list.


In [101]:
# to look at target_cities_notin_massachusettes_zipcodes:
target_cities_notin_massachusetts_usazip

['Alford',
 'Aquinnah',
 'Bourne',
 'Clarksburg',
 'Dartmouth',
 'Easton',
 'Egremont',
 'Florida',
 'Foxborough',
 'Freetown',
 'Gosnold',
 'Hamilton',
 'Hancock',
 'Hardwick',
 'Hawley',
 'Lanesborough',
 'Leyden',
 'Middleborough',
 'Middlefield',
 'Monroe',
 'Montgomery',
 'Mount Washington',
 'New Ashford',
 'New Marlborough',
 'Pelham',
 'Peru',
 'Phillipston',
 'Shelburne',
 'Stockbridge',
 'Tisbury',
 'Tolland',
 'Truro',
 'Tyngsborough',
 'Tyringham',
 'Warren',
 'Washington',
 'West Tisbury',
 'Westhampton',
 'Whately',
 'Yarmouth',
 'Unknown']

In [111]:
massachusetts_usazip_notin_target_cities = []

for city in list(massachusetts_usazip['Major_city']):
    if city in list(target_cities):
        print(city)
    if city not in list(target_cities):
        print('>', city)
        massachusetts_usazip_notin_target_cities.append(city)

Agawam
Amherst
Barre
Belchertown
Blandford
Brimfield
Chester
Chesterfield
Chicopee
Chicopee
Chicopee
Cummington
Easthampton
East Longmeadow
> Feeding Hills
> Gilbertville
Goshen
Granby
Granville
Hadley
Hampden
Hatfield
> Haydenville
Holyoke
Huntington
> Leeds
Leverett
Ludlow
Monson
Northampton
> Florence
Oakham
Palmer
Plainfield
Russell
Shutesbury
Southampton
South Hadley
Southwick
> Three Rivers
Wales
Ware
> West Chesterfield
Westfield
> West Hatfield
West Springfield
Wilbraham
Williamsburg
Worthington
Springfield
Springfield
Springfield
Longmeadow
Springfield
Springfield
Springfield
Springfield
Springfield
Springfield
Springfield
Springfield
> Indian Orchard
Springfield
Springfield
Pittsfield
Adams
> Ashley Falls
Becket
> Berkshire
Cheshire
Dalton
Great Barrington
Hinsdale
> Housatonic
> Lanesboro
Lee
Lenox
Monterey
North Adams
Otis
Richmond
Sandisfield
Savoy
Sheffield
> South Egremont
> Southfield
West Stockbridge
Williamstown
Windsor
Greenfield
Ashfield
Athol
Bernardston
Buckland
C

In [117]:
print('There are', str(len(massachusetts_usazip_notin_target_cities)), 'massachusetts usazip cities not in the target cities list.')       

There are 111 massachusetts usazip cities not in the target cities list.


### There are a variety of reasons why these lists do not match up
- Many of the zipcodes in Boston have their own 'common names' (e.g. Roxbury). In the USAZipcode database 'Boston' is listed as a "common_city_name" for Roxbury. These type of issues are easily resolved either by scripting (cycling through the common_city_name) or by manually inspecting (what I actually did).
- Other discrepencies are very easy (e.g. Yarmouth vs Yarmouth Port; Foxboro vs Foxborough).
- Most of the discrepencies are in rural areas of Massachusettes. These type of discrepencies fall into a few categories:
- (1) Small towns adjacent to each other are both reported in the massachusettes overdose data (good) but share the same zipcode (bad). This issue could be resolved by summing the city overdose data based on zipcode to get a composite city (e.g. Athol and Phillipson are reported as seperate towns by MA, but they share the same zipcode). 
- (2) Some rural regions are difficult to resolve. The NE and SE corners of Massachusettes have many small towns that may or may not be reported, and the zipcodes can be disjunct and overlapping.

### An example of why zip codes are challenging...

- Alford is in the Massachusettes government overdose data 
- Alford is not in the zipcode databases
- Alford has two zipcodes according to google search: 01230 and 01266

In [118]:
# Alford is next to Great Berrington 
# But is not in the common_city_list
massachusetts_usazip[massachusetts_usazip['Major_city']=='Great Barrington']

Zipcode Zipcode_type        Major_city      Post_office_city  \
71   01230     Standard  Great Barrington  Great Barrington, MA   

                                     Common_city_list            County State  \
71  [Great Barrington, Egremont, Gt Barrington, N ...  Berkshire County    MA   

     Lat   Lng Timezone  ...  Land_area_in_sqmi Water_area_in_sqmi  \
71  42.2 -73.3  Eastern  ...              91.52               1.75   

    Housing_units  Occupied_housing_units  Median_home_value  \
71         4633.0                  3479.0           351100.0   

    Median_household_income  Bounds_west  Bounds_east  Bounds_north  \
71                  51467.0   -73.476856   -73.180617     42.278609   

    Bounds_south  
71     42.087676  

[1 rows x 24 columns]

In [119]:
# Alford is next to Stockbridge / West Stockbridge 
# And is in the common_city_list
massachusetts_usazip[massachusetts_usazip['Major_city']=='West Stockbridge']

Zipcode Zipcode_type        Major_city      Post_office_city  \
86   01266     Standard  West Stockbridge  West Stockbridge, MA   

                             Common_city_list            County State    Lat  \
86  [West Stockbridge, Alford, W Stockbridge]  Berkshire County    MA  42.29   

      Lng Timezone  ...  Land_area_in_sqmi Water_area_in_sqmi  Housing_units  \
86 -73.37  Eastern  ...              26.05                0.3         1096.0   

    Occupied_housing_units  Median_home_value  Median_household_income  \
86                   675.0           400000.0                  76953.0   

    Bounds_west  Bounds_east  Bounds_north  Bounds_south  
86   -73.440786   -73.312309     42.352016     42.225038  

[1 rows x 24 columns]

- Alford is in West Stockbridge

In [121]:
massachusetts_city_overdoses[massachusetts_city_overdoses['City'] == "West Stockbridge"]

City 2014 2015 2016 2017 2018
325  West Stockbridge    0    1    0    0    0

- Alford is a city in the Mass overdose data
- West Stockbridge is a city in Mass overdose AND in the Zip dictionary
- So the demographic data at Zip code 01266 will go towards West Stockbridge
- But will not go towards Alford. 
- So a sensical thing to do would be to merge Alford OD data into West Stockbridge OD data...
- For now, if I just use "cities in both" I will be removing Alford OD but any data associated with ZIP code 01266 will be for Alford + West Stockbridge
- I think this would result in an underestimation of overdose because:
- (ODs in W Stockbridge / (population size of Alford + W stockbridge) is less than the true amount

# In the end... here is my final zipcode dictionary!
- 303 unique target cities (out of 352 reported by Massachusetts)

In [151]:
zipcode_dictionary = pd.read_csv('~/Desktop/new_insights/data_files/final_zipcode_dictionary.csv', dtype={'Zipcode':str}, index_col = False)


In [152]:
temp = pd.merge(zipcode_dictionary, target_cities_df, on = 'City')
unique_cities = temp['City'].unique()
len(unique_cities)

303